File: persistence/src/main/java/io/sparkled/persistence/playlist/impl/query/GetSequenceAtPlaylistIndexQuery.java
package io.sparkled.persistence.playlist.impl.query;

import io.sparkled.model.entity.Sequence;
import io.sparkled.model.entity.Stage;
import io.sparkled.persistence.PersistenceQuery;
import io.sparkled.persistence.QueryFactory;

import java.util.Optional;

public class GetSequenceAtPlaylistIndexQuery implements PersistenceQuery<Optional<Sequence>> {

    private final int playlistId;
    private final int index;

    public GetSequenceAtPlaylistIndexQuery(int playlistId, int index) {
        this.playlistId = playlistId;
        this.index = index;
    }

    @Override
    public Optional<Sequence> perform(QueryFactory queryFactory) {
        final Sequence sequence = queryFactory
                .selectFrom(qSequence)
                .innerJoin(qPlaylistSequence).on(qSequence.id.eq(qPlaylistSequence.sequenceId))
                .where(qPlaylistSequence.playlistId.eq(playlistId))
                .orderBy(qPlaylistSequence.displayOrder.asc())
                .offset(index)
                .limit(1)
                .fetchFirst();

        return Optional.ofNullable(sequence);
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/playlist/impl/query/GetSequenceAtPlaylistIndexQuery.kt
package io.sparkled.persistence.playlist.impl.query

import io.sparkled.model.entity.QPlaylistSequence.Companion.playlistSequence
import io.sparkled.model.entity.QSequence.Companion.sequence
import io.sparkled.model.entity.Sequence
import io.sparkled.persistence.PersistenceQuery
import io.sparkled.persistence.QueryFactory

class GetSequenceAtPlaylistIndexQuery(private val playlistId: Int, private val index: Int) :
    PersistenceQuery<Sequence?> {

    override fun perform(queryFactory: QueryFactory): Sequence? {
        return queryFactory
            .selectFrom(sequence)
            .innerJoin(playlistSequence).on(sequence.id.eq(playlistSequence.sequenceId))
            .where(playlistSequence.playlistId.eq(playlistId))
            .orderBy(playlistSequence.displayOrder.asc())
            .offset(index.toLong())
            .limit(1)
            .fetchFirst()
    }
}
